tools/managed/mcm_with_timelock.go

```diff
@@ -0,0 +1,137 @@
+package managed
+
+import (
+	"math/big"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/rlp"
+	"github.com/smartcontractkit/ccip-owner-contracts/tools/errors"
+	"github.com/smartcontractkit/ccip-owner-contracts/tools/executable"
+	owner "github.com/smartcontractkit/ccip-owner-contracts/tools/gethwrappers"
+)
+
+const (
+	ZERO_HASH = "0x0000000000000000000000000000000000000000000000000000000000000000"
+)
+
+type MCMSWithTimelockChainMetadata struct {
+	executable.ExecutableMCMSChainMetadata
+	TimelockAddress common.Address `json:"timelockAddress"`
+}
+
+type TimelockOperation string
+
+const (
+	Schedule TimelockOperation = "schedule"
+	Cancel   TimelockOperation = "cancel"
+	Bypass   TimelockOperation = "bypass"
+)
+
+type MCMSWithTimelockProposal struct {
+	baseMCMSProposal
+
+	Operation TimelockOperation `json:"operation"` // Always 'schedule', 'cancel', or 'bypass'
+
+	// TODO: this should be configurable as a human-readable string
+	// i.e. 1d, 1w, 1m, 1y
+	MinDelay string `json:"minDelay"`
+
+	// Map of chain identifier to chain metadata
+	ChainMetadata map[string]MCMSWithTimelockChainMetadata `json:"chainMetadata"`
+
+	// Operations to be executed
+	Transactions []DetailedBatchChainOperation `json:"transactions"`
+}
+
+func (m *MCMSWithTimelockProposal) Validate() error {
+	if err := m.baseMCMSProposal.Validate(); err != nil {
+		return err
+	}
+
+	switch m.Operation {
+	case Schedule, Cancel, Bypass:
+		break
+	default:
+		return &errors.ErrInvalidTimelockOperation{
+			ReceivedTimelockOperation: string(m.Operation),
+		}
+	}
+
+	_, err := time.ParseDuration(m.MinDelay)
+	if err != nil {
+		return &errors.ErrInvalidMinDelay{
+			ReceivedMinDelay: m.MinDelay,
+		}
+	}
+
+	return nil
+}
+
+func (m *MCMSWithTimelockProposal) ToExecutableMCMSProposal() (executable.ExecutableMCMSProposal, error) {
```

Comment:
This laying of embedding is pretty hard to follow and I could imagine it leading to fields empty unintentionally/difficult initialization. Can we invert the design a bit and just have constructors for the 2 cases we have here? Example
```
package mcms

import (
	"crypto"
	"encoding/binary"
	"fmt"
	"math/big"
)

type TimelockBatch struct {
	Ops []Op `json:"ops"`
	Description string `json:"description"`
}

type Op struct {
	ChainIdentifer string         `json:"chain_identifer,omitempty"`
	To             common.Address `json:"to,omitempty"`
	Value          big.Int        `json:"value"`
	Data           []byte         `json:"data,omitempty"`
}

type Proposal struct {
	Version              string      `json:"version"`
	ValidUntil           uint32      `json:"validUntil"`
	Signatures           []Signature `json:"signatures"`
	OverridePreviousRoot bool        `json:"overridePreviousRoot"`

	// Map of chain identifier to chain metadata
	ChainMetadata map[string]ExecutableMCMSChainMetadata `json:"chainMetadata"`
	Operations []Op `json:"ops"`
}


func (p Proposal) Sign(signer func(h [32]byte) (Signature, error), clients map[string]bind.ContractBackend) error {
	root, err := generateRoot(clients, p)
	if err != nil {
		return err
	}
	var validUntilBytes [32]byte
	binary.BigEndian.PutUint32(validUntilBytes[28:], p.ValidUntil) // Place the uint32 in the last 4 bytes
	hashToSign := crypto.Keccak256Hash(root, validUntilBytes[:])
	sig, err := signer(toEthSignedMessageHash(hashToSign))
	if err != nil {
		return err
	}
	p.Signatures = append(p.Signatures, sig)
	return nil	
}

type ChainMetadata struct {
	NonceOffset uint64         `json:"nonceOffset" :"nonce_offset"`
	MCMAddress  common.Address `json:"mcmAddress" :"mcm_address"`
}

func NewProposal(
	version string, // Type?
	validUntil uint32, // Type?
	overridePreviousRoot bool,
	chainMetadata map[string]ExecutableMCMSChainMetadata,
	operations []Op,
) (Proposal, error) {
	// Validate parameters
	return Proposal{
	}, nil
}

func NewProposalFromTimelockBatches(
	batches []TimelockBatch,
	version string, 
	validUntil uint32, // Type?
	overridePreviousRoot bool,
	chainMetadata map[string]ExecutableMCMSChainMetadata,
	operations []Op,
	) (Proposal, error) {
	// Validate parameters
	return Proposal{
		
	}, nil
}
```
I'm not sure we need the executor abstraction either with above approach

Reply:
Hm let me think about this for a bit, but in the meantime, How would we construct a proposal from a file in this case then? Because signatures will be getting aggregated through some file, so we need to be able to unmarshall the JSON into a given proposal structure

Reply:
> This laying of embedding is pretty hard to follow and I could imagine it leading to fields empty unintentionally/difficult initialization.
Not following why this is the case - validations ensure all required fields exist, and this is the only entrypoint that consumers need to concern themselves with.